src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -21706,6 +21706,451 @@ void M(string? s)
                 );
         }
 
+        // TODO2
+        // test on indexer
+
+        [Fact]
+        public void NotNullIfNotNull_Return()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(""p"")] public string? M(string? p) => throw null!;
+}";
+            var source = @"
+class D
+{
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        c.M(p1).ToString();
+        c.M(p2).ToString(); // 1
+    }
+}";
+            var expected = new[]
+            {
+                // (8,9): warning CS8602: Dereference of a possibly null reference.
+                //         c.M(p2).ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.M(p2)").WithLocation(8, 9)
+            };
+
+            var lib = CreateNullableCompilation(new[] { NotNullIfNotNulltAttributeDefinition, lib_cs });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(expected);
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(expected);
+            CompileAndVerify(comp2);
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_Return_NullableInt()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(""p"")] public int? M(int? p) => throw null!;
+}";
+            var source = @"
+class D
+{
+    static void M(C c, int? p1, int? p2)
+    {
+        _ = p1 ?? throw null!;
+        c.M(p1).Value.ToString();
+        c.M(p2).Value.ToString(); // 1
+    }
+}";
+            var expected = new[]
+            {
+                // (8,9): warning CS8629: Nullable value type may be null.
+                //         c.M(p2).Value.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "c.M(p2)").WithLocation(8, 9)
+            };
+
+            var lib = CreateNullableCompilation(new[] { NotNullIfNotNulltAttributeDefinition, lib_cs });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(expected);
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(expected);
+            CompileAndVerify(comp2);
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_Return_MultipleParameters()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(""p""), NotNullIfNotNull(""p2"")] public string? M(string? p, string? p2) => throw null!;
```

Comment:
So this is supposed to be an OR relationship? Is there any way to annotate an AND relationship?

Reply:
Correct.
If `p` is not null, then the return is not null.
If `p2` is not null, then the return is not null.
This is the behavior discussed in LDM.

There is no way to do an AND.

---
In reply to: [303560635](https://github.com/dotnet/roslyn/pull/37201#discussion_r303560635) [](ancestors = 303560635)